feat(travel): skeleton create Travel

minyong96 opened the pull request:
## Summary
- create travel 불변식 조건과 스켈레톤 코드 작성
- snapshot (readModel ) 정의 
- EntityGraph 사용
- 스냅샷 정상작동 테스트 코드작성
- command 추가
- 제약조건 추가

## Related Issue
- Issue Number: #21 

## Changes
- requiredContructor로 변경
- List 불변
- TravelCommand 추가 
- 테스트로직 작성 

## Discussion Topic
- roadMap은 읽기전용모델로 반환되는건가? 아니면 엔티티로 반환되는건가
- 내가 현재 roadMap에서 필요한 부분이 무엇인가? 읽기모델로 가져다 쓰면, 성능이슈가 나올거같은데 roadMap에서 dto로 주는게 더 나은 방법같긴한데.. 고민

## Checklist
- [ ] 로컬 테스트 완료
- [x] lint 통과(로컬)
- [x] 리뷰어 요청(최소 2명), 라벨/스코프 지정
- [ ] 브랜치 최신화(rebase) 및 충돌 없음

minyong96 committed:
feat(travel): skeleton create Travel

dd-jiyun commented on src/main/java/com/kllhy/roadmap/travel/domain/model/ProgressSubTopic.java:
```diff
@@ -16,5 +21,20 @@ public class ProgressSubTopic extends IdAuditEntity {
+        this.subTopicId = Objects.requireNonNull(subTopicId, "subTopicId must not be null");
+    }
+
+    static ProgressSubTopic create(ProgressTopic topic, Long subTopicId) {
+        return new ProgressSubTopic(topic, subTopicId);
+    }
+
+    void setTopic(ProgressTopic topic) {
+        this.topic = Objects.requireNonNull(topic, "topic must not be null");
+    }
+
+    void changeStatus(ProgressStatus status) {
+        this.status = status;
+    }
```
해당 메서드들 접근 제어자가 `default`로 되어있는데 이유가 있으신가요?

minyong96 replied:
해당 메서드는 같은 패키지에 있는 root만 메서드를 사용할수있어서 (외부 패키지 접근불가)
default로 해놨습니당 👍

dd-jiyun commented on src/main/java/com/kllhy/roadmap/travel/domain/model/ProgressTopic.java:
```diff
@@ -23,7 +31,59 @@ public class ProgressTopic extends IdAuditEntity {
+    void addSubTopics(List<ProgressSubTopicCommand> commands) {
+        if (commands == null || commands.isEmpty()) return;
+
+        var ids = commands.stream().map(ProgressSubTopicCommand::subTopicId).toList();
```
ProgressSubTopicCommand에서 가지고 있는 값이 `Long subTopicId`인데 var 타입으로 두신 이유가 궁금합니다!
제가 알기로 var 타입은 타입을 추론하는데 이 과정에서 추론에 실패할 경우 예외가 발생하는 것으로 알고 있어서요!

minyong96 replied:
오 그런가요? 지역변수 내에서만 사용하고 버릴거라고 가독성 측면에서 var로 해놨는데
근데 지금 ids 를 참조하고있는부분이 없네용? 제가 불변조건 검사를 제거해놔서 필요없는 구문이긴한데 
말씀하신대로 타입추론은 
ProgressTopicCommand라고 명확하게 정의해놔서 괜찮지 않을까싶습니다!
감사합니당!

dd-jiyun reviewed (commented):
고생하셨습니다!

youabledev commented on src/main/java/com/kllhy/roadmap/travel/domain/model/Travel.java:
```diff
@@ -22,7 +30,57 @@ public class Travel extends AggregateRoot {
+    }
+
+    public void markTopic(Long topicId, ProgressStatus status) {
+        Objects.requireNonNull(status, "status");
```
status 값이 널인 경우 즉시 NPE가 발생하게 되는데 의도하신건지
아니면 추후 CustomException으로 정의하실지 궁금합니다!

minyong96 replied:
넵 의도했습니다.
사실 당연히 status 가 들어와야되는건데 개발자의 실수를 방지하고자 내부가드를 둔거라서요
이부분은 exception을 예외를 던지면 더 찾기 어려워져서 NPE로 발생하게끔해놨습니당

youabledev commented on src/main/java/com/kllhy/roadmap/travel/domain/model/command/ProgressTopicCommand.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.Objects;
+
+public record ProgressTopicCommand(Long topicId, List<ProgressSubTopicCommand> subTopics) {
+    public ProgressTopicCommand {
```
도메인 생성시 필요한 값을 정의하기 위해 사용하면서, 
Command 객체 생성 시점에 값의 유효성을 검증할 수 있어서 좋은 것 같습니다!

minyong96 committed:
fix: addSubTopics of ProgressTopic

youabledev commented on src/main/java/com/kllhy/roadmap/travel/domain/model/command/ProgressTopicCommand.java:
```diff
@@ -0,0 +1,19 @@
+        if (topicId == null) throw new DomainException(TravelErrorCode.TRAVEL_TOPICS_INVALID);
+
+        if (subTopics == null) {
+            subTopics = List.of();
```
record에 기본 값을 지정할 수 없어서 subTopics의 값이 null 인 경우 빈 배열로 지정해 주신 것 좋습니다! 👍

youabledev commented on src/main/java/com/kllhy/roadmap/travel/domain/model/ProgressSubTopic.java:
```diff
@@ -3,11 +3,16 @@
+    @Getter private Long subTopicId;
 
     @Enumerated(EnumType.STRING)
     private ProgressStatus status = ProgressStatus.TODO;
```
enum 기본 값으로 TODO가 설정된다면 해당 컬럼은 nullable false일 것 같은데
@Column(nullable=false)로 명시해 놓는 거 어떠신가요??!

youabledev reviewed (approved):
고생하셨습니다! 
올려주신 코드 보면서 프로젝트 구조에 대한 감을 잡을 수 있었습니다!

minyong96 committed:
fix: nullable false

SeungYeop-Han commented on src/main/java/com/kllhy/roadmap/travel/application/command/TravelServiceAdapter.java:
```diff
@@ -1,8 +1,21 @@
+
+    @Override
+    public void create(Long userId, Long roadmapId) {
+        // user와 roadmap 찾아와서
```
내일

논의했던 내용대로 AggregateRoot 인 RoadMap 엔티티에 대한 리드 모델 정의한 후, .toReadModel 메서드 통해서 변환하도록 처리하겠습니다.

그리고 application 패키지의 comman 쪽에서 받아보실 수 있도록 인터페이스 만들어 놓겠습니다

minyong96 replied:
제가 올린 PR에 저는 스냅샷(readModel 용도) 형태로 만들어뒀습니당
스냅샷이 필요할때랑 아닐때랑을 구분하여 필요시에는 가볍게 ( 빠른 데이터 조회, 성능이슈) dto로 정의하여 바로 repository에서 projection으로 가져올 생각입니당
제가 필요한 부분 스펙만 정의해주면 projection으로 view 반환도 괜찮은 방법같은데 한번 고민해보는것도 괜찮을거같습니다!

SeungYeop-Han commented on src/main/java/com/kllhy/roadmap/travel/domain/model/ProgressSubTopic.java:
```diff
@@ -3,18 +3,39 @@
+import lombok.NoArgsConstructor;
 
 @Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
이런식으로 jpa 에서 필요로하는 기본 생성자를 정의하면 누락 위험을 줄일 수 있을 것 같습니다.

좋은 방법인 것 같습니다!

SeungYeop-Han commented on src/main/java/com/kllhy/roadmap/travel/domain/model/ProgressTopic.java:
```diff
@@ -1,29 +1,88 @@
     private Travel travel;
 
-    private Long topicId;
+    @Getter private Long topicId;
```
이 부분 뿐만 아니라, progress topic 및 progress sub topic 에 대한 필드 및 메서드 이름에 progress 를 빼신 이유가 코드 간결함 때문인가요?

저는 개인적으로 붙여주는 쪽을 선호하긴 합니다. 민영님 의견 들어보고 싶습니다

minyong96 replied:
이 클래스가 속한 Aggregate 내에서는 progress라는 접두어가 중복 정보이기 때문에 
Travel 과 관련된 토픽임을 이미 내포하고 있다고 생각해서 제거했습니당

**DDD의 장점은 유비쿼터스 언어잖아용!**

minyong96 committed:
feat: add snapshot

- apply db constraint
- add snapshot Test
- apply entityGraph

# Conflicts:
#	src/test/resources/application.yaml

minyong96 committed:
refactor: add TravelCommand

SeungYeop-Han reviewed (approved):
수고 많으셨습니다!

minyong96 committed:
refactor: remove snapshot and feat travelView
